ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+
+One example comes from [`IPC::MicroSocket::Server`](https://metacpan.org/pod/IPC::MicroSocket::Server). This is currently using the `Sublike::Extended` module (see "Prototype Implementation" below) to provide the syntax. The actual module also uses `Object::Pad` to provide the `method` keyword; this example is paraphrased to avoid that. This module requires a `path` named argument, and optionally takes a `listen` argument, defaulting its value to 5 if the caller did not provide a defined value.
+
+```perl
+extended sub new_unix ( $class, :$path, :$listen //= 5 )
+{
+    ...
+}
+```
+
+This replaces the previous version of the code which handled arguments by the more traditional approach of assigning into a hash:
+
+```perl
+sub new_unix ( $class, %args )
+{
+    my $path   = $args{path};
+    my $listen = $args{listen} // 5;
+    ...
+}
+```
+
+Already the new code is shorter and more consise. Additionally it contains error checking that complains about missing mandatory keys, or unrecognised keys, which the previous version of the code did not include.
+
+Another example, this time from [`Text::Treesitter::QueryCursor`](https://metacpan.org/pod/Text::Treesitter::QueryCursor). This method takes an optional argument named `multi`, tested for truth. If absent it should default to false.
+
+```perl
+sub next_match_captures ( $self, :$multi = 0 )
+{
+    ...
+}
+```
+
+The previous version of this code did include complaints about unrecognised keys, and was rather longer because of it:
+
+```perl
+sub next_match_captures ( $self, %options )
+{
+    my $multi = delete $options{multi};
+    keys %options and
+        croak "Unrecognised options to ->next_captures: " . join( ", ", keys %options );
+
+    ...
+}
+```
+
+## Prototype Implementation
+
+The [`XS::Parse::Sublike`](https://metacpan.org/pod/XS::Parse::Sublike) module contains parsing to allow third-party syntax modules to parse subroutine-like constructions, and includes a parser for named parameters already having this syntax. These are also made available to regular subroutines via the `extended` keyword provided by [`Sublike::Extended`](https://metacpan.org/pod/Sublike::Extended).
+
+## Future Scope
+
+* If the Metaprogramming API (PPC0022) gains introspection abilities to enquire about subroutine signature parameters, further consideration will need to be made in that API on how to represent the extra kinds of parameters added by this specification.
+
+## Rejected Ideas
+
+* This specification intentionally makes no changes to the call-site of any subroutines using named parameters.
+
+## Open Issues
+
+* How to specify a named parameter whose name is anything other than the name of the lexical variable into which its value is assigned? This is more of note when considering that traditional argument handling techniques involving assignment into hashes can handle more styles of name that would be invalid for lexical variables - for example, names including hyphens.
```

Comment:
Another issue is that subroutines can't currently check whether named parameters with default values had something passed in by the caller. That is, if you do
```perl
sub foo(:$arg => 42) {
    ...
}
```
then the two calls `foo()` and `foo(arg => 42)` are indistinguishable by `foo`.

Reply:
Imean... Yes, that's true, but if such a distinction is actually meaningful to the function then I'd suggest "don't do that". If you need to distinguish a 42 being given from the caller vs 42 from your own default, then you need to rethink something. Maybe just take an undef default, check it for being defined into some flag, and then `//= 42` it anyway..?

Reply:
I could think of some use-cases, for example code that inserts something to a database. `undef` is `NULL`, while no value means "insert default value".

Of course this could be handled manually, with a slurpy hash parameter.

Hmm... what if exists() could be used on named arguments?

Reply:
PS. I agree it's a niche case and it's probably not worth worrying about.

Reply:
@xenu you can still start to use `%_` to store only arguments set by caller so to distinguish set value and default value you can easily do:
```
if (exists $_{arg})
```

Reply:
While ugly, if we don't want to support some special syntax for this (I think) rare cae, the function could set the default to an unlikely value and check for that:
```
my $unknown;
# x can be supplied as any value, but \$unknown isn't accessible elsewhere
sub f ( :$x = \$unknown) { 
  if (ref $x && $x == \$unknown) {
    # whatever we want to do that's special
  }
}
```